app/Models/MplsSap.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+use LibreNMS\Interfaces\Models\Keyable;
+
+class MplsSap extends Model implements Keyable
+{
+    protected $primaryKey = 'sap_id';
+    public $timestamps = false;
+    protected $fillable = [
+        'svc_oid',
+        'sapPortId',
+        'ifName',
+        'sapEncapValue',
+        'device_id',
+        'sapRowStatus',
+        'sapType',
+        'sapDescription',
+        'sapAdminStatus',
+        'sapOperStatus',
+        'sapLastMgmtChange',
+        'sapLastStatusChange',
+    ];
+
+    // ---- Helper Functions ----
+
+    /**
+     * Get a string that can identify a unique instance of this model
+     * @return string
+     */
+    public function getCompositeKey()
+    {
+        return $this->svc_oid . '-' . $this->sapPortId . '-' . $this->sapEncapValue;
+    }
+
+    // ---- Define Relationships ----
+
+    public function binds()
+    {
+        return $this->hasMany('App\Models\MplsSdpBind', 'svc_oid');
+    }
+    
+    public function services()
+    {
+        return $this->hasMany('App\Models\MplsService', 'svc_oid');
```

Comment:
This relationship won't work because svc_oid is not unique across all devices.
You need to store the unique id of the service.  (Same with binds)

Reply:
I confess that I don't really understand for what are object relationships needed, just copied that part from app/Models/MplsLsp.php :-o
can I just omit the binds() and services() methods? I do not see that these were used somewhere.

otherwise getting and store the keys would make the polling and discovery code more complex.

Reply:
no, because it isn't the code that is wrong, it is the database table structure.  Yes, it does make it more complex in the lsp code, I passed the parent objects to the child polling code so they wouldn't have to be re-fetched from the db.

Reply:
I have updated the code but something is wrong with unit tests now.
--- Expected
+++ Actual
@@ @@
     'mpls_sdps' => Array (...)
     'mpls_sdp_binds' => Array (
         0 => Array (
-            'sdp_id' => 19
-            'svc_id' => 13
+            'sdp_id' => 2
+            'svc_id' => 1

everitime I run the
./scripts/save-test-data.php -o timos -m mpls
the Script generate new ids.
On my development environment I dont see any problems in the database tables. The Ids are consistent.